sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+// directly navigate to the page.
+cy.visit('/endpoint');
+
+// to ensure stability, wait for network (preferred), ui elements, or both
+
+// note: you can optionally add a url check
+cy.url().should('contain', 'endpoint');
+// cy.url().should('match', /endpoint/); // there are many, some more complex, ways of doing it
+
+
+// network wait:
+cy.server();  // required to use cy.route
+// usually a GET request. Is aliased so we can wait for it.
+cy.route('some-xhr-call-that-happens-upon-landing').as('crutcXHR');
+cy.wait('@crutchXHR', {timeout: 15000}); // can add an optional timeout if the page takes too long
+
+// ui-element wait is straightforward, and may be optional, as well as less stable)
+cy.get('element-on-page').should('exist').and('be.visible');
+
+```
+
+<br/><br/>
+
+### Network stub record & play
+This is an advanced technique that strongly relates to UI-integration tests. Recall UI-integration references [1](../testing-strategy/component-vs-integration-vs-e2e-testing.md), [2](../real-life-examples/test-front-end-with-integration-back-end-with-e2e.md).
+
+Cypress allows to stub all network traffic. We can record the network data from an endpoint, and stub that response every time the UI makes a call to an arbitrary server.
+
+Start by copying the network data from devTools to a json file. Place it in `cypress/fixtures` folder. This folder is made for this purpose, and any reference to it will default to the root of the folder.
+
+![devtools > network tab](../../assets/images/ui-state/devtools-network.PNG)
+
+```javascript
+// json has been copied to a file `cypress/fixtures/agents.json`
+
+// this is a shorthand for cy.fixture(). More at https://docs.cypress.io/api/commands/fixture.html#Accessing-Fixture-Data
+cy.server();
+cy.route('some-xhr-call-that-happens-upon-landing', 'fx:agents.json').as('crutcXHR');
+// all calls to the network route will be stubbed by the data in agents.json file
+```
+
+#### What if there are so many network requests happening?
+Where do we get all our mocks for fixtures? We do not want to manually copy and save them. We want to record them as the test runs against a real API
+
+There are at least 2 Cypress plugins you can utilize for this [1](https://github.com/Nanciee/cypress-autorecord) & [2](https://github.com/scottschafer/cypressautomocker).
+
+If these do not work for you, you can easily create record and playback utilities yourself with these 3 functions.
+
+```javascript
+function stubRecorder(pathToJson) {
+  const xhrData = []; // an empty array to hold the data
+  cy.server({ // if recording, save the response data in the array
+    onResponse: (response) => {
+      const url = response.url;
+      const method = response.method;
+      const data = response.response.body;
+      // We push a new entry into the xhrData array
+      xhrData.push({ url, method, data });
+    }
+  });
+  // this is used as a ‘filter’ for the data you want to record.
```

Comment:
I have not understood what `this` is, I do not see anything that looks like a filter

Reply:
added clarifying comments